cdap-security/src/main/java/co/cask/cdap/security/authorization/AuthorizationEnforcementService.java

```diff
@@ -0,0 +1,154 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.security.authorization;
+
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.proto.id.EntityId;
+import co.cask.cdap.proto.security.Action;
+import co.cask.cdap.proto.security.Principal;
+import co.cask.cdap.proto.security.Privilege;
+import co.cask.cdap.security.spi.authorization.AuthorizationEnforcer;
+import co.cask.cdap.security.spi.authorization.Authorizer;
+import co.cask.cdap.security.spi.authorization.UnauthorizedException;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.AbstractScheduledService;
+import com.google.inject.Inject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * A {@link AuthorizationEnforcer} used to enforce authorization policies in programs. If caching is enabled using
+ * {@link co.cask.cdap.common.conf.Constants.Security.Authorization#CACHE_ENABLED}, authorization policies are cached
+ * locally. A thread refreshes the cached policies periodically.
+ */
+public class AuthorizationEnforcementService extends AbstractScheduledService implements AuthorizationEnforcer {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AuthorizationEnforcementService.class);
+
+  private final AuthorizerInstantiator authorizerInstantiator;
+  private final boolean authorizationEnabled;
+  private final boolean cacheEnabled;
+  private final int cacheTtlSeconds;
+  private final Map<Principal, Set<Privilege>> authPolicyCache;
+  private final Lock readLock;
+  private final Lock writeLock;
+
+  @Inject
+  public AuthorizationEnforcementService(AuthorizerInstantiator authorizerInstantiator, CConfiguration cConf) {
+    this.authorizerInstantiator = authorizerInstantiator;
+    this.authorizationEnabled = cConf.getBoolean(Constants.Security.Authorization.ENABLED);
+    this.cacheEnabled = cConf.getBoolean(Constants.Security.Authorization.CACHE_ENABLED);
+    this.cacheTtlSeconds = cConf.getInt(Constants.Security.Authorization.CACHE_TTL_SECS);
+    this.authPolicyCache = new HashMap<>();
+    ReadWriteLock lock = new ReentrantReadWriteLock();
+    this.readLock = lock.readLock();
+    this.writeLock = lock.writeLock();
+  }
+
+  @Override
+  protected Scheduler scheduler() {
+    return Scheduler.newFixedDelaySchedule(0, cacheTtlSeconds, TimeUnit.SECONDS);
+  }
+
+  @Override
+  protected void runOneIteration() throws Exception {
+    if (!cacheEnabled) {
```

Comment:
I tried to use an approach which created multiple ``AuthorizationEnforcementService`` classes that get injected based on cConf here - https://github.com/caskdata/cdap/commit/96dd11b377d45f0457763c10c2ed01be19ea1d61. However, it just seems to me that at this stage, it adds a lot of almost no-op code. Not sure if there is much value to doing that yet. 

I have however added a check for authorizationEnabled here just to cut the point across, as well as a warning in the ``startUp`` method if authorization is disabled but caching is enabled.